core/trino-main/src/main/java/io/trino/execution/scheduler/BinPackingNodeAllocatorService.java

```diff
@@ -0,0 +1,435 @@
+            InternalNodeManager nodeManager,
+            ClusterMemoryManager clusterMemoryManager,
+            NodeSchedulerConfig config)
+    {
+        this(nodeManager,
+                requireNonNull(clusterMemoryManager, "clusterMemoryManager is null")::getWorkerMemoryInfo,
+                config.isIncludeCoordinator());
+    }
+
+    @VisibleForTesting
+    BinPackingNodeAllocatorService(
+            InternalNodeManager nodeManager,
+            Supplier<Map<String, Optional<MemoryInfo>>> workerMemoryInfoSupplier,
+            boolean scheduleOnCoordinator)
+    {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.workerMemoryInfoSupplier = requireNonNull(workerMemoryInfoSupplier, "workerMemoryInfoSupplier is null");
+        this.scheduleOnCoordinator = scheduleOnCoordinator;
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        if (started.compareAndSet(false, true)) {
+            executor.schedule(() -> {
+                while (!stopped.get()) {
+                    try {
+                        // pending acquires are processed when node is released (semaphore is bumped) and periodically (every couple seconds)
+                        // in case node list in cluster have changed.
+                        processSemaphore.tryAcquire(PROCESS_PENDING_ACQUIRES_DELAY_SECONDS, TimeUnit.SECONDS);
+                        processSemaphore.drainPermits();
+                        processPendingAcquires();
+                    }
+                    catch (InterruptedException e) {
+                        currentThread().interrupt();
+                    }
+                    catch (Exception e) {
+                        // ignore to avoid getting unscheduled
+                        log.warn(e, "Error updating nodes");
+                    }
+                }
+            }, 0, TimeUnit.SECONDS);
+        }
+
+        refreshNodePoolSizes();
+        executor.scheduleWithFixedDelay(this::refreshNodePoolSizes, 1, 5, TimeUnit.SECONDS);
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        stopped.set(true);
+        executor.shutdownNow();
+    }
+
+    private void refreshNodePoolSizes()
+    {
+        ImmutableMap.Builder<String, Long> newNodePoolSizes = ImmutableMap.builder();
+
+        Map<String, Optional<MemoryInfo>> workerMemoryInfo = workerMemoryInfoSupplier.get();
+        for (Map.Entry<String, Optional<MemoryInfo>> entry : workerMemoryInfo.entrySet()) {
+            if (entry.getValue().isEmpty()) {
+                continue;
+            }
+            newNodePoolSizes.put(entry.getKey(), entry.getValue().get().getPool().getMaxBytes());
+        }
+        nodePoolSizes.set(newNodePoolSizes.buildOrThrow());
+    }
+
+    @VisibleForTesting
+    synchronized void processPendingAcquires()
+    {
+        // synchronized only for sake manual triggering in test code. In production code it should only be called by single thread
```

Comment:
But even in test, we call it single-threadedly, right?

Reply:
The thread which calls out to `processPendingAcquires` is still running in tests so we need synchronized section here. I could disable the thread in tests but I would rather keep test specific changes in class minimal. (synchronized section feels not a big deal to me)